src/generic/GenericTypes.jl

```diff
@@ -798,6 +798,35 @@ end
 
 const PuiseuxSeriesElem{T} = Union{PuiseuxSeriesRingElem{T}, PuiseuxSeriesFieldElem{T}} where T <: RingElement
 
+###############################################################################
+#
+#   AbsMSeriesRing / AbsMSeries
+#
+###############################################################################
+
+mutable struct AbsMSeriesRing{T <: RingElement} <: AbstractAlgebra.MSeriesRing{T}
+   base_ring::Ring
+   poly_ring::AbstractAlgebra.MPolyRing{T}
+   prec_max::Vector{Int}
+   sym::Vector{Symbol}
+
+   function AbsMSeriesRing{T}(R::Ring, poly_ring::AbstractAlgebra.MPolyRing{T}, prec::Vector{Int}, s::Vector{Symbol}, cached::Bool = true) where T <: RingElement
+      return get_cached!(AbsMSeriesID, (R, prec, s), cached) do
+         new{T}(R, poly_ring, prec, s)
+      end::AbsMSeriesRing{T}
+   end
+end
+
+const AbsMSeriesID = CacheDictType{Tuple{Ring, Vector{Int}, Vector{Symbol}}, Ring}()
+
+mutable struct AbsMSeries{T <: RingElement} <: AbstractAlgebra.AbsMSeriesElem{T}
+   poly::AbstractAlgebra.MPolyElem{T}
+   prec::Vector{Int}
+   parent::AbsMSeriesRing{T}
```

Comment:
```suggestion
mutable struct AbsMSeries{S, T <: RingElement} <: AbstractAlgebra.AbsMSeriesElem{T}
   poly::S
   prec::Vector{Int}
   parent::AbsMSeriesRing{T}
```
Is there a reason to not parametrize with the polynomial type and the coefficient type? The current implementation is type unstable in every function call that uses `.poly`.

Reply:
I can change that if you like.

Reply:
I think that would increase perfomance greatly for things using the series. But we can also change it later.